artifactory/services/utils/multipartupload.go

```diff
@@ -0,0 +1,459 @@
+		}
+	}()
+
+	if err = mu.uploadPartsConcurrently(logMsgPrefix, fileSize, splitCount, localPath, progressReader, multipartUploadClient); err != nil {
+		return
+	}
+
+	if sha1 == "" {
+		sha1, err = calculateSha1(localPath)
+		if err != nil {
+			return
+		}
+	}
+
+	if progress != nil {
+		progressReader = progress.SetMergingState(progressReader.GetId(), false)
+	}
+
+	log.Info(logMsgPrefix + "Starting merging parts...")
+	// The total number of attempts is determined by the number of retries + 1
+	return mu.completeAndPollForStatus(logMsgPrefix, mu.retries+1, sha1, multipartUploadClient, progressReader)
+}
+
+func (mu *MultipartUpload) uploadPartsConcurrently(logMsgPrefix string, fileSize int64, splitCount int, localPath string, progressReader ioutils.Progress, multipartUploadClient *httputils.HttpClientDetails) (err error) {
+	numberOfParts := calculateNumberOfParts(fileSize)
+	log.Info(fmt.Sprintf("%sSplitting file to %d parts, using %d working threads for uploading...", logMsgPrefix, numberOfParts, splitCount))
+	producerConsumer := parallel.NewRunner(splitCount, uint(numberOfParts), false)
+
+	wg := new(sync.WaitGroup)
+	wg.Add(int(numberOfParts))
+	attemptsAllowed := new(atomic.Uint64)
+	attemptsAllowed.Add(uint64(numberOfParts) * uint64(mu.retries))
+	go func() {
+		for i := 0; i < int(numberOfParts); i++ {
+			if err = mu.produceUploadTask(producerConsumer, logMsgPrefix, localPath, fileSize, numberOfParts, int64(i), progressReader, multipartUploadClient, attemptsAllowed, wg); err != nil {
+				return
+			}
+		}
+	}()
+	go func() {
+		defer producerConsumer.Done()
+		wg.Wait()
+	}()
+	producerConsumer.Run()
+	if attemptsAllowed.Load() == 0 {
+		return errorutils.CheckError(errTooManyAttempts)
+	}
+	return nil
+}
+
+func (mu *MultipartUpload) produceUploadTask(producerConsumer parallel.Runner, logMsgPrefix, localPath string, fileSize, numberOfParts, partId int64, progressReader ioutils.Progress, multipartUploadClient *httputils.HttpClientDetails, attemptsAllowed *atomic.Uint64, wg *sync.WaitGroup) (retErr error) {
+	_, taskErr := producerConsumer.AddTaskWithError(func(int) error {
+		uploadErr := mu.uploadPart(logMsgPrefix, localPath, fileSize, partId, progressReader, multipartUploadClient)
+		if uploadErr == nil {
+			log.Info(fmt.Sprintf("%sCompleted uploading part %d/%d", logMsgPrefix, partId+1, numberOfParts))
+			wg.Done()
+		}
+		return uploadErr
+	}, func(uploadErr error) {
+		if attemptsAllowed.Load() == 0 {
+			wg.Done()
+			return
+		}
+		log.Warn(fmt.Sprintf("%sPart %d/%d - %s", logMsgPrefix, partId+1, numberOfParts, uploadErr.Error()))
+		attemptsAllowed.Add(^uint64(0))
+
+		// Sleep before trying again
+		time.Sleep(retriesInterval)
+		err := mu.produceUploadTask(producerConsumer, logMsgPrefix, localPath, fileSize, numberOfParts, partId, progressReader, multipartUploadClient, attemptsAllowed, wg)
+		if err != nil {
+			retErr = err
+		}
+	})
+	if taskErr != nil {
+		retErr = taskErr
+	}
```

Comment:
Do we actually need this if statement? Can we instead always set the value of `retErr` to `taskErr `?
Or maybe even better than that -  use `retErr` instead of  `taskErr` in the first line of this function - 
```
_, taskErr := producerConsumer.AddTaskWithError(func(int) error {
```